20241016-R-error-code-metrics.md

```diff
@@ -0,0 +1,108 @@
+
+When opted-in, in tandem with the current behavior, where the errors are logged, a new metric `dapr_error_code_count` would provide the # of times an error code has been thrown for each `app-id`. This opt-in is independent from the `debug` log mode and thus can be used separately.
+
+## Background
+
+Currently, error codes are only visible as logs when Dapr runs in `debug` log mode, which is impractical for both production apps and observability. 
+
+Additionally, the error message is a field in JSON response body as such:
+```
+{
+	...
+    "errorCode": "ERR_STATE_GET",
+    ...
+}
+```
+...which means observation of these codes requires increased log noise via the `debug` flag, as well as parsing the logs to find this field. There also wouldn't be automatic data retention, so logs would have to be manually stored to provide a historical view.
+
+
+Dapr uses Prometheus by default, and so having these errors as Prometheus metrics will allow querying, aggregation, and data retention of the codes.
+
+## Related Items
+
+Initial Error Codes Proposal: [proposals/#21](https://github.com/dapr/proposals/pull/21)
+
+Error Codes Implementation Parent: [dapr/#6068](https://github.com/dapr/dapr/issues/6068)
+
+## Expectations and alternatives
+
+### Expectations
+- When opted-in, it would add up to  `N * app`  (N = number of error codes) metrics for scraping, today that's around `52 * app`.
+- If a component-based error, codes could repeat for multiple components used by the same app, and wouldn't be differentiated by the metrics, which only provide the `app-id`.
+- Many logged error codes are currently defined in [pkg/messages/predefined.go](https://github.com/dapr/dapr/blob/master/pkg/messages/predefined.go), but not all. Implementation requires case-by-case analysis of usage of error codes in that uppercase format `ERR_NAME_TAG`.
+
+### Alternatives
+- One alternative solution which retains error codes as logs only would be to change the level of logs containing these error codes from `debug` to `info`. However, this increases noise and raises logs which may be unhelpful at `info` level.
+- Another alternative solution would be to log error codes separately with opt-in functionality, but this achieves the same result as adding metrics, but increase log noise and doesn't introduce a simple way to actually work with the errors.
+
+## Implementation Details
+
+### Solution
+
+The feature is enabled via an opt-in flag within Dapr HTTP metrics (disabled by default). The goal of the opt-in functionality is to prevent the increased cost of additional metrics if they aren't being used.
+
+```yaml
+spec:
+  metric:
+    enabled: true
+    http:
+      errorCodeMetrics: true
+```
+From the user's perspective, the metric `dapr_error_code_count` could be split by `app-id` and `error_code` like so:
+```go
+dapr_error_code_count{app_id="service-a", error_code="ERR_ACTOR_RUNTIME_NOT_FOUND"} 4
```

Comment:
Also, to what interval are those counts emitted? I would include the timestamp and the time window too. The time window should probably be configurable too, so users might decide that they need it every 1min, 5min, 10min, 1h, depending on the app's needs.

Reply:
Indeed, metric granularity is based on the scrape interval. Prometheus is a time series database so metrics are associated with a timestamp naturally. Users tune metric collection outside of Dapr on the Prometheus side.